packages/client/src/bags/types/bagOperation.types.ts

```diff
@@ -0,0 +1,177 @@
+  variantId?: OperationChange['variantId'];
+  oldValue?: OperationChange['oldValue'];
+  newValue?: OperationChange['newValue'];
+};
+
+type ChangeTypeItemOriginChanged = {
+  changeType: ChangeType.ItemOriginChanged;
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+};
+
+type ChangeTypeItemPriceChanged = {
+  changeType: ChangeType.ItemPriceChanged;
+  itemId?: OperationChange['itemId'];
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+};
+
+type ChangeTypeItemQuantityChanged = {
+  changeType: ChangeType.ItemQuantityChanged;
+  itemId?: OperationChange['itemId'];
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+};
+
+type ChangeTypeItemNotEnoughStock = {
+  changeType: ChangeType.NotEnoughStock;
+  itemId?: OperationChange['itemId'];
+  stockAvailable?: OperationChange['stockAvailable'];
+};
+
+type ChangeTypePromoCodeEligibilityChanged = {
+  changeType: ChangeType.PromoCodeEligibilityChanged;
+};
+
+type ChangeTypePromotionNotApplied = {
+  changeType: ChangeType.PromotionNotApplied;
+};
+
+type ChangeTypeShippingCostChanged = {
+  changeType: ChangeType.ShippingCostChanged;
+};
+
+type ChangeTypeShippingFlatRateEligibilityChanged = {
+  changeType: ChangeType.ShippingFlatRateEligibilityChanged;
+};
+
+type ChangeTypeTotalDiscountChanged = {
+  changeType: ChangeType.TotalDiscountChanged;
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+};
+
+type ChangeTypeSelectedSaleIntentChanged = {
+  changeType: ChangeType.SelectedSaleIntentChanged;
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+  itemId?: OperationChange['itemId'];
+  oldValue?: OperationChange['oldValue'];
+  newValue?: OperationChange['newValue'];
+};
+
+export type BagOperationChange =
+  | ChangeTypeItemDiscountChanged
+  | ChangeTypeItemIsNowAvailable
+  | ChangeTypeItemIsUnavailable
+  | ChangeTypeItemMerged
+  | ChangeTypeItemOriginChanged
+  | ChangeTypeItemPriceChanged
+  | ChangeTypeItemQuantityChanged
+  | ChangeTypeItemNotEnoughStock
+  | ChangeTypePromoCodeEligibilityChanged
+  | ChangeTypePromotionNotApplied
+  | ChangeTypeSelectedSaleIntentChanged
+  | ChangeTypeShippingCostChanged
+  | ChangeTypeShippingFlatRateEligibilityChanged
+  | ChangeTypeTotalDiscountChanged;
+
+export enum BagViolationFixSuggestionAction {
+  DecreaseQuantity = 'DecreaseQuantity',
+  IncreaseValue = 'IncreaseValue',
+  DecreaseValue = 'DecreaseValue',
+}
+
+type BagViolationFixSuggestion = {
+  action: BagViolationFixSuggestionAction;
+  value: string;
+};
+
+export type BagViolation = {
+  itemsIds?: Array<BagItem['id']>;
+  fixSuggestion?: BagViolationFixSuggestion;
+};
+
+export type BagOperation = {
+  id: string;
+  createdDate: string;
+  changes: Array<BagOperationChange>;
```

Comment:
This is more of a question just because I'm finding it hard to understand - shouldn't this be `Array<OperationChange>`? I mean, if the operation is a `ChangeTypeItemDiscountChanged`, won't we receive all properties anyway, instead of the 4 you have defined for that type? 🤔 
Looking at the API, the "related properties" for each type feel purely informative, just for the developer to know what to look for in a given scenario, but all properties are received as well. I don't know, just raising a question 🙂

Reply:
The other properties are not received :) Only the related ones.